File: sdk/javascript/recipient.js
import crypto from "node:crypto";
import * as jose from 'jose';
import {
  JWKSCache,
  fetchJWKSWithRetry,
  findSigningKey,
  getKeyType,
  getAlgorithmForKey,
  publicKeyToJWK
} from "./utils.js";
import { getRecipientUserAgent } from './version.js';

/**
 * TACRecipient - Implements the recipient side of the Trusted Agentic Commerce Protocol
 * Handles TAC-Protocol message verification and decryption
 */
class TACRecipient {
  /**
   * @param {Object} options - Configuration options
   * @param {string} options.domain - Domain of the recipient (required)
   * @param {crypto.KeyObject|string} options.privateKey - Private key for decryption - RSA or EC (required)
   * @param {number} options.cacheTimeout - JWKS cache timeout in ms (default: 3600000)
   * @param {number} options.maxRetries - Max retry attempts for network requests (default: 3)
   * @param {number} options.retryDelay - Retry delay in ms (default: 1000)
   */
  constructor(options = {}) {
    if (!options.domain) {
      throw new Error("domain is required in TACRecipient constructor");
    }
    if (!options.privateKey) {
      throw new Error("privateKey is required in TACRecipient constructor");
    }
    
    this.domain = options.domain;
    this.setPrivateKey(options.privateKey); // This sets both private and public keys
    this.jwksCache = new JWKSCache(options.cacheTimeout || 3600000);
    this.maxRetries = options.maxRetries || 3;
    this.retryDelay = options.retryDelay || 1000;
  }

  /**
   * Set private key and automatically derive public key
   * @param {crypto.KeyObject|string} privateKey - Private key object or PEM string (RSA or EC)
   * @private
   */
  setPrivateKey(privateKey) {
    if (typeof privateKey === "string") {
      this.privateKey = crypto.createPrivateKey(privateKey);
    } else {
      this.privateKey = privateKey;
    }

    // Verify it's a supported key type
    const supportedTypes = ['rsa', 'rsa-pss', 'ec'];
    if (!supportedTypes.includes(this.privateKey.asymmetricKeyType)) {
      throw new Error('TAC Protocol requires RSA or EC (P-256/384/521) keys');
    }

    // Always derive public key from private key
    this.publicKey = crypto.createPublicKey(this.privateKey);
    
    // Store key type and algorithm
    this.keyType = getKeyType(this.privateKey);
    this.encryptionAlgorithm = getAlgorithmForKey(this.privateKey, 'enc');
  }

  /**
   * Generate key ID from public key
   * @returns {string} Base64url encoded SHA-256 hash of public key
   * @throws {Error} If no public key is available
   */
  generateKeyId() {
    if (!this.publicKey) {
      throw new Error("No public key available. Load or set keys first.");
    }
    const keyData = this.publicKey.export({ type: 'spki', format: 'der' });
    return crypto.createHash('sha256').update(keyData).digest('base64url');
  }

  /**
   * Fetch JWKS from a domain's well-known endpoint
   * @param {string} domain - Domain to fetch JWKS from
   * @param {boolean} forceRefresh - Force cache bypass
   * @returns {Promise<Array>} Array of JWK objects
   */
  async fetchJWKS(domain, forceRefresh = false) {
    return await fetchJWKSWithRetry(domain, {
      cache: this.jwksCache,
      maxRetries: this.maxRetries,
      initialDelay: this.retryDelay,
      maxDelay: this.retryDelay * 30,
      userAgent: getRecipientUserAgent(),
      forceRefresh,
    });
  }

  /**
   * Process TAC-Protocol message
   * @param {string} tacMessage - Base64-encoded TAC-Protocol message (from header or body)
   * @returns {Promise<Object>} Processing result with valid, issuer, expires, data, recipients, errors
   */
  async processTACMessage(tacMessage) {
    const result = {
      valid: false,
      issuer: null,
      expires: null,
      data: null,
      recipients: [],
      errors: [],
    };

    if (!tacMessage) {
      result.errors.push('Missing TAC-Protocol message');
      return result;
    }

    // Decode base64 message
    let decodedMessage;
    try {
      // Try to decode as base64 first
      decodedMessage = Buffer.from(tacMessage, 'base64').toString('utf8');
      // Validate it's valid JSON
      JSON.parse(decodedMessage);
    } catch (e) {
      // If base64 decode fails, assume it's already JSON (backward compatibility)
      decodedMessage = tacMessage;
    }

    try {
      // Parse the multi-recipient message
      let message;
      try {
        message = JSON.parse(decodedMessage);
      } catch (e) {
        throw new Error('Invalid TAC-Protocol message format');
      }

      // Validate message structure
      if (!message.recipients || !Array.isArray(message.recipients)) {
        throw new Error('Invalid message format: missing recipients');
      }

      result.recipients = message.recipients.map(r => r.kid || 'unknown');

      // Find our specific JWE
      const ourRecipient = message.recipients.find(r => r.kid === this.domain);
      if (!ourRecipient) {
        throw new Error(`Not a recipient: ${this.domain}`);
      }

      if (!this.privateKey) {
        throw new Error('No private key available for decryption');
      }

      // Decrypt and verify our specific JWE in one step
      // jose.jwtDecrypt automatically verifies the JWT signature during decryption
      const { payload } = await jose.jwtDecrypt(ourRecipient.jwe, this.privateKey);
      
      // Verify the issuer and audience claims
      if (!payload.iss) {
        throw new Error('JWT missing issuer (iss) claim');
      }
      
      if (payload.aud !== this.domain) {
        throw new Error(`JWT audience mismatch: expected ${this.domain}, got ${payload.aud}`);
      }

      // Additional signature verification by fetching sender's public key
      const agentDomain = payload.iss;
      const jwks = await this.fetchJWKS(agentDomain);
      
      if (!jwks || jwks.length === 0) {
        throw new Error(`No public keys found for agent ${agentDomain}`);
      }

      // Find appropriate signing key to verify this came from the right sender
      const signingKey = findSigningKey(jwks);
      if (!signingKey) {
        throw new Error('No suitable signing key found');
      }
      
      // The payload is already verified through JWE decryption, but we could
      // add additional verification here if needed for extra security

      // Extract data from verified payload
      result.valid = true;
      result.issuer = payload.iss;
      result.expires = new Date(payload.exp * 1000);
      
      // Get data specific to this recipient
      result.data = payload.data || null;

    } catch (error) {
      result.errors.push(error.message);
    }

    return result;
  }

  /**
   * Static method to inspect TAC-Protocol message without decryption
   * @param {string} tacMessage - Base64-encoded TAC-Protocol message
   * @returns {Object} Basic information about the message
   */
  static inspect(tacMessage) {
    try {
      // Decode base64 message
      let decodedMessage;
      try {
        decodedMessage = Buffer.from(tacMessage, 'base64').toString('utf8');
        JSON.parse(decodedMessage);
      } catch (e) {
        decodedMessage = tacMessage;
      }
      
      const message = JSON.parse(decodedMessage);
      
      return {
        version: message.version || '2025-08-21',
        recipients: message.recipients ? message.recipients.map(r => r.kid || 'unknown') : [],
        expires: null // Cannot determine without decryption
      };
    } catch (error) {
      return {
        error: 'Invalid TAC-Protocol message format',
        recipients: []
      };
    }
  }

  /**
   * Get public key as JWK for publishing (bidirectional use)
   * @returns {Promise<Object>} JWK representation of the public key
   * @throws {Error} If no public key is available
   */
  async getPublicJWK() {
    return await publicKeyToJWK(this.publicKey, this.generateKeyId());
  }

  /**
   * Clear JWKS cache
   * @param {string} domain - Specific domain to clear, or null for all
   */
  clearCache(domain = null) {
    this.jwksCache.clear(domain);
  }
}

export default TACRecipient;

File: sdk/python/src/recipient.py
"""
TACRecipient - Implements the recipient side of the Trusted Agentic Commerce Protocol
Handles TAC-Protocol message verification and decryption
"""

import base64
import hashlib
import json
import re
import time
from typing import Any, Dict, List, Optional, Union

import jose.jwt
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from jose import jwe
from jose.exceptions import JWTError

try:
    from .errors import TACCryptoError, TACErrorCodes, TACMessageError, TACNetworkError, TACValidationError
    from .utils import (
        JWKSCache,
        fetch_jwks_with_retry,
        find_signing_key,
        get_algorithm_for_key,
        get_key_type,
        get_user_agent,
        public_key_to_jwk,
    )
except ImportError:
    from errors import TACCryptoError, TACErrorCodes, TACMessageError, TACNetworkError, TACValidationError
    from utils import (
        JWKSCache,
        fetch_jwks_with_retry,
        find_signing_key,
        get_algorithm_for_key,
        get_key_type,
        get_user_agent,
        public_key_to_jwk,
    )


class TACRecipient:
    """
    TACRecipient - Implements the recipient side of the Trusted Agentic Commerce Protocol
    Handles TAC-Protocol message verification and decryption
    """

    def __init__(
        self,
        domain: str,
        private_key: Union[str, Any],
        cache_timeout: int = 3600000,
        max_retries: int = 3,
        retry_delay: int = 1000,
        clock_tolerance: int = 300,
        password: Optional[bytes] = None,
        hide_user_agent_version: bool = False,
    ):
        """
        Initialize TACRecipient

        Args:
            domain: Domain of the recipient (required)
            private_key: Private key for decryption (required)
            cache_timeout: JWKS cache timeout in ms (default: 3600000)
            max_retries: Max retry attempts for network requests (default: 3)
            retry_delay: Retry delay in ms (default: 1000)
            clock_tolerance: Clock skew tolerance in seconds (default: 300 = 5 minutes)
            password: Password for encrypted private keys (default: None)
            hide_user_agent_version: If True, omit version details from User-Agent header (default: False)
        """
        if not domain:
            raise TACValidationError("domain is required in TACRecipient constructor", TACErrorCodes.DOMAIN_REQUIRED)
        if not private_key:
            raise TACValidationError(
                "privateKey is required in TACRecipient constructor", TACErrorCodes.PRIVATE_KEY_REQUIRED
            )

        self.domain = domain
        self._password = password
        self.set_private_key(private_key)  # This sets both private and public keys
        self.jwks_cache = JWKSCache(cache_timeout)
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.clock_tolerance = clock_tolerance
        self.hide_user_agent_version = hide_user_agent_version

    def set_private_key(self, private_key: Union[str, Any], password: Optional[bytes] = None):
        """
        Set private key and automatically derive public key

        Args:
            private_key: Private key object or PEM string
            password: Password for encrypted private keys (default: uses constructor password)
        """
        key_password = password if password is not None else getattr(self, "_password", None)
        try:
            if isinstance(private_key, str):
                self.private_key = serialization.load_pem_private_key(private_key.encode(), password=key_password)
            else:
                self.private_key = private_key
        except Exception as error:
            raise TACCryptoError(f"Invalid key data: {str(error)}", TACErrorCodes.INVALID_KEY_DATA)

        # Verify it's a supported key type
        if not isinstance(self.private_key, rsa.RSAPrivateKey):
            raise TACCryptoError(
                "TAC Protocol requires RSA keys (minimum 2048-bit, 3072-bit recommended)",
                TACErrorCodes.UNSUPPORTED_KEY_TYPE,
            )

        # Always derive public key from private key
        self.public_key = self.private_key.public_key()

        # Store key type and algorithm
        self.key_type = get_key_type(self.private_key)
        self.encryption_algorithm = get_algorithm_for_key(self.private_key, "enc")

    def generate_key_id(self) -> str:
        """
        Generate key ID from public key

        Returns:
            Base64url encoded SHA-256 hash of public key
        """
        if not self.public_key:
            raise TACValidationError("No public key available. Load or set keys first.", TACErrorCodes.NO_PUBLIC_KEY)

        key_data = self.public_key.public_bytes(
            encoding=serialization.Encoding.DER, format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        return base64.urlsafe_b64encode(hashlib.sha256(key_data).digest()).decode().rstrip("=")

    async def fetch_jwks(self, domain: str, force_refresh: bool = False) -> List[Dict]:
        """
        Fetch JWKS from a domain's well-known endpoint

        Args:
            domain: Domain to fetch JWKS from
            force_refresh: Force cache bypass

        Returns:
            Array of JWK objects
        """
        return await fetch_jwks_with_retry(
            domain,
            cache=self.jwks_cache,
            max_retries=self.max_retries,
            retry_delay=self.retry_delay,
            max_delay=self.retry_delay * 30,
            user_agent=get_user_agent(hide_version=self.hide_user_agent_version),
            force_refresh=force_refresh,
        )

    async def process_tac_message(self, tac_message: Optional[str]) -> Dict[str, Any]:
        """
        Process TAC-Protocol message

        Args:
            tac_message: Base64-encoded TAC-Protocol message (from header or body)

        Returns:
            Processing result with valid, issuer, expires, data, recipients, errors
        """
        result = {
            "valid": False,
            "issuer": None,
            "expires": None,
            "jti": None,  # JWT ID for replay detection
            "data": None,
            "recipients": [],
            "errors": [],
        }

        if not tac_message:
            result["errors"].append("Missing TAC-Protocol message")
            return result

        # Check message size limit (100KB max to prevent DoS)
        MAX_MESSAGE_SIZE = 100 * 1024  # 100KB
        if len(tac_message) > MAX_MESSAGE_SIZE:
            result["errors"].append(f"Message too large: {len(tac_message)} bytes exceeds maximum of {MAX_MESSAGE_SIZE} bytes")
            return result

        # Decode base64 message (strictly required - raw JSON not accepted)
        # Check if input looks like raw JSON (not base64 encoded)
        trimmed = tac_message.strip()
        if trimmed.startswith("{") or trimmed.startswith("["):
            result["errors"].append("Invalid TAC-Protocol message: must be base64-encoded (raw JSON not accepted)")
            return result

        # Validate base64 format
        base64_regex = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")
        if not base64_regex.match(tac_message):
            result["errors"].append("Invalid TAC-Protocol message: must be base64-encoded")
            return result

        try:
            decoded_message = base64.b64decode(tac_message).decode("utf-8")
        except Exception:
            result["errors"].append("Invalid TAC-Protocol message: must be base64-encoded")
            return result

        try:
            # Parse the multi-recipient message
            try:
                message = json.loads(decoded_message)
            except Exception:
                raise TACMessageError("Invalid TAC-Protocol message format", TACErrorCodes.INVALID_MESSAGE_FORMAT)

            # Validate message structure
            if "recipients" not in message or not isinstance(message["recipients"], list):
                raise TACMessageError("Invalid message format: missing recipients", TACErrorCodes.MISSING_RECIPIENTS)

            result["recipients"] = [r.get("kid", "unknown") for r in message["recipients"]]

            # Find our specific JWE
            our_recipient = None
            for recipient in message["recipients"]:
                if recipient.get("kid") == self.domain:
                    our_recipient = recipient
                    break

            if not our_recipient:
                raise TACMessageError(f"Not a recipient: {self.domain}", TACErrorCodes.NOT_A_RECIPIENT)

            if not self.private_key:
                raise TACValidationError("No private key available for decryption", TACErrorCodes.NO_PRIVATE_KEY)

            # Step 1: DECRYPT the JWE to get the signed JWT
            try:
                # Use the private key directly for decryption - jose library will handle the key type
                decrypted_jwt = jwe.decrypt(our_recipient["jwe"], self.private_key)
                signed_jwt = decrypted_jwt.decode("utf-8")
            except Exception as error:
                raise TACCryptoError(f"Decryption failed: {str(error)}", TACErrorCodes.DECRYPTION_FAILED)

            # Step 2: VERIFY the JWT signature using sender's public key
            # First, get the sender's domain and key ID from the JWT (without verification)
            try:
                unverified_header = jose.jwt.get_unverified_header(signed_jwt)
                unverified_payload = jose.jwt.get_unverified_claims(signed_jwt)
            except Exception as error:
                raise TACCryptoError(f"JWT decode failed: {str(error)}", TACErrorCodes.JWT_DECODE_FAILED)

            if "iss" not in unverified_payload:
                raise TACMessageError("JWT missing issuer (iss) claim", TACErrorCodes.JWT_MISSING_ISSUER)

            if unverified_payload.get("aud") != self.domain:
                raise TACMessageError(
                    f'JWT audience mismatch: expected {self.domain}, got {unverified_payload.get("aud")}',
                    TACErrorCodes.JWT_AUDIENCE_MISMATCH,
                )

            # Step 3: Fetch sender's public key and VERIFY the JWT signature
            agent_domain = unverified_payload["iss"]
            key_id = unverified_header.get("kid")  # Extract key ID for key selection

            try:
                jwks_keys = await self.fetch_jwks(agent_domain)
            except Exception as error:
                raise TACNetworkError(
                    f"Failed to fetch JWKS for {agent_domain}: {str(error)}", TACErrorCodes.JWKS_FETCH_FAILED
                )

            if not jwks_keys:
                raise TACNetworkError(
                    f"No public keys found for agent {agent_domain}", TACErrorCodes.NO_PUBLIC_KEYS_FOUND
                )

            # Find appropriate signing key to verify the JWT signature (using key ID if available)
            signing_key = find_signing_key(jwks_keys, key_id)
            if not signing_key:
                raise TACNetworkError("No suitable signing key found", TACErrorCodes.NO_SIGNING_KEY_FOUND)

            try:
                # VERIFY the JWT signature - this is the crucial security step!
                payload = jose.jwt.decode(
                    signed_jwt,
                    signing_key,
                    algorithms=[signing_key.get("alg", "RS256")],
                    audience=self.domain,
                    issuer=agent_domain,
                )
            except JWTError as error:
                raise TACCryptoError(
                    f"Signature verification failed: {str(error)}", TACErrorCodes.SIGNATURE_VERIFICATION_FAILED
                )

            # Additional manual validation for iat claim (not validated by jose by default)
            now = int(time.time())

            if payload.get("iat") and payload["iat"] > now + self.clock_tolerance:
                raise TACMessageError("JWT not yet valid (issued in the future)", TACErrorCodes.JWT_NOT_YET_VALID)

            # Extract data from verified payload
            result["valid"] = True
            result["issuer"] = payload.get("iss")
            result["expires"] = time.gmtime(payload.get("exp", 0)) if payload.get("exp") else None
            result["jti"] = payload.get("jti")  # JWT ID for replay detection

            # Get data specific to this recipient
            result["data"] = payload.get("data")

        except Exception as error:
            result["errors"].append(str(error))

        return result

    @staticmethod
    def inspect(tac_message: str) -> Dict[str, Any]:
        """
        Static method to inspect TAC-Protocol message without decryption

        Args:
            tac_message: Base64-encoded TAC-Protocol message

        Returns:
            Basic information about the message
        """
        try:
            import re

            # Check if input looks like raw JSON (not base64 encoded)
            trimmed = tac_message.strip()
            if trimmed.startswith("{") or trimmed.startswith("["):
                return {
                    "error": "Invalid TAC-Protocol message: must be base64-encoded (raw JSON not accepted)",
                    "version": None,
                    "recipients": [],
                    "expires": None,
                }

            # Validate base64 format
            base64_regex = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")
            if not base64_regex.match(tac_message):
                return {
                    "error": "Invalid TAC-Protocol message: must be base64-encoded",
                    "version": None,
                    "recipients": [],
                    "expires": None,
                }

            # Decode base64 message
            decoded_message = base64.b64decode(tac_message).decode("utf-8")
            message = json.loads(decoded_message)

            return {
                "version": message.get("version", "2025-08-27"),
                "recipients": [r.get("kid", "unknown") for r in message.get("recipients", [])],
                "expires": None,  # Cannot get expiry without decryption
            }
        except Exception:
            return {"error": "Invalid TAC-Protocol message format", "version": None, "recipients": [], "expires": None}

    async def get_public_jwk(self) -> Dict[str, Any]:
        """
        Get public key as JWK for publishing (bidirectional use)

        Returns:
            JWK representation of the public key
        """
        return public_key_to_jwk(self.public_key, self.generate_key_id())

    def clear_cache(self, domain: Optional[str] = None):
        """
        Clear JWKS cache

        Args:
            domain: Domain to clear cache for, or None to clear all
        """
        self.jwks_cache.clear(domain)
